Moved most of the cookie and network interaction to the injected page

berndsi opened the pull request:
When "keep me logged in" was not used during log in, the session cookie
is not accessible in the global page. By moving the network request to
fetch the CSRF token to the injected/content page (popups.js) this
works now.

So, the structure is more like this:
* global page: for UI + I18N
* injected page: network requests + DOM handling of current wiki page

Between those messages have to be dispatched. Care has to be taken that
only simple, cloneable objects are passed with the message as message
data.

Bug: T191998

berndsi committed:
Moved most of the cookie and network interaction to the injected page

When "keep me logged in" was not used during log in, the session cookie
is not accessible in the global page. By moving the network request to
fetch the CSRF token to the injected/content page (popups.js) this
works now.

So, the structure is more like this:
* global page: for UI + I18N
* injected page: network requests + DOM handling of current wiki page

Between those messages have to be dispatched. Care has to be taken that
only simple, cloneable objects are passed with the message as message
data.

Bug: T191998

berndsi committed:
No instance variable for canonicalPageTitle anymore

It's probably safer not to deal with instance variables.
Although having moved it to the injected page was probably an
improvement already. But with the new structure we don't need it.

berndsi committed:
Stringify error object

To avoid getting errors like '[object Object]'.

mdholloway commented:
This PR allows me to save exactly one page in the case where I don't check "keep me logged in" and have "prevent cross-site tracking" checked in the privacy settings.  Then after clicking a link I'm logged out.  And if I have "prevent cross-site tracking" unchecked, then I can keep navigating and saving pages.  So we're on the right track!  Nice find on those docs/writeups.

I wonder if the remaining issues are because of the i18n message fetch that's still happening in `popup.js`, in the function `geti18nMessages`.

berndsi commented:
@mdholloway I can add multiple pages to the reading list without any issues. Can you try again?
If that still happens you can also add a `console.log(token);` inside the `handleTokenResult` function.
I don't think the i18n messages need to be moved to the injected.js as well. They don't deal with cookies AFAIU.

berndsi commented:
@mdholloway I have the same settings you mentioned: "keep me logged in" is off and have "prevent cross-site tracking" is on. I can add multiple pages to the default reading list without issues. Are there any other Safari settings you changed? If I turn on "Block all cookies" I cannot even log in, so that can't be it.

mdholloway reviewed (changes requested):
I tested again, and I am still logged out after saving the first page in the scenario we've been discussing.  However, if I change the value of the `credentials` option from `same-origin` to `omit` in the fetch call in `geti18nMessages`, then the patch works without any issues.  With that change I think this is good to merge.

berndsi committed:
Omit credentials for I18N messages request

Providing the credentials from popup.js causes the user being logged
out sometimes.

berndsi committed:
Bump version to 0.3

mdholloway reviewed (commented):
I think CFBundleVersion should be bumped to '3' as well.

berndsi commented:
@mdholloway Both ways work for me. Thanks! I was testing with Safari 11.1 on High Sierra.

mdholloway commented on WikipediaReadingLists.safariextension/Info.plist:
```diff
@@ -13,7 +13,7 @@
-	<string>0.2</string>
+	<string>0.3</string>
 	<key>CFBundleVersion</key>
 	<string>2</string>
```
Sorry, that last comment pertains to this value.  (Not used to reviewing on GitHub...)

berndsi replied:
Done. Sorry, was wondering about that but wasn't sure.

mdholloway commented:
Weird, Version 11.1 (13605.1.33.1.4) on High Sierra here, too.  We'll have to make sure both Rita and Anthony test it out again before releasing.

berndsi committed:
Also bump CFBundleVersion

tgr commented:
Without credentials, the messages won't be in the user's language.

mdholloway commented:
@tgr  ...Yeah, you're right.  I double-checked for sanity and I am still inappropriately logged out in the usual case if cookies are sent.  I'm working on an update now to move the message fetching over to the injected script, which should resolve this once and for all.
